modules/layers/src/solid-polygon-layer/solid-polygon-layer-vertex-side.glsl.js

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2015 - 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import main from './solid-polygon-layer-vertex-main.glsl';
+
+export default `\
+#define SHADER_NAME solid-polygon-layer-vertex-shader-side
+#define IS_SIDE_VERTEX
+
+attribute vec2 vertexPositions;
+
+attribute vec3 instancePositions;
+attribute vec2 instancePositions64xyLow;
+attribute vec3 nextPositions;
+attribute vec2 nextPositions64xyLow;
+attribute float instanceElevations;
+attribute vec4 instanceFillColors;
+attribute vec4 instanceLineColors;
+attribute vec3 instancePickingColors;
+attribute float vertexValid;
+
+uniform bool extruded;
+uniform bool isWireframe;
+uniform float elevationScale;
+uniform float opacity;
+
+varying vec4 vColor;
+varying float isValid;
+
+void main(void) {
+  vec3 positions = instancePositions;
+  vec2 positions64xyLow = instancePositions64xyLow;
+  float elevations = instanceElevations;
+  vec4 fillColors = instanceFillColors;
+  vec4 lineColors = instanceLineColors;
+  vec3 pickingColors = instancePickingColors;
+  ${main}
```

Comment:
Done this way, `main` works more like a C MACRO, i.e. unstructured textual replacement.

If instead, you defined a function in `main`, you could call the function here.

```
varying vec4 vColor;
varying float isValid;

{main}

void main(void) {
  params.position = ..
  ...
  calculatePosition(params);
}
```


The downside is the parameter passing, but main could then also define the uniforms it needed etc.

You would essentially be on your way to defining a local shader module. Which is completely feasible, as they essentially work exactly like this.

Reply:
Ah, neat trick. I like how this keeps all relevant data (structs, attributes, function) in one place.